File: crawlers/ai/extract.py
"""
Provider-agnostic structured-extraction wrapper for crawlers.

Routes JSON-extraction calls to either Anthropic Haiku or OpenAI gpt-4o-mini
based on the `provider` argument, enforces per-run budget caps via
`crawler_runs.cost_usd`, parses + schema-validates the response, and returns
a Python dict (or raises). Every call's cost is recorded to the run row.

Usage:

    from crawlers.ai.extract import extract_json, BudgetExceeded

    data = extract_json(
        prompt="...",
        run_id=run.run_id,
        max_usd_per_run=5.0,
        provider="openai",      # or "anthropic"
        expected_keys=("title", "scholarships"),
    )

Provider routing default: 'openai' (gpt-4o-mini) for production extraction —
4-5x cheaper per token than Haiku.

NOTE: this wrapper is intentionally small and stateless. It does NOT cache
prompts, does NOT batch — those are agent-level optimisations the Discoverer
will layer on top.
"""

import json
import os
import re
from typing import Iterable, Optional

import httpx
from dotenv import load_dotenv

load_dotenv(dotenv_path=os.path.join(os.path.dirname(__file__), "..", "..", ".env.local"))

SB_URL = os.environ["NEXT_PUBLIC_SUPABASE_URL"]
SB_KEY = os.environ["SUPABASE_SERVICE_ROLE_KEY"]
SB_H = {"apikey": SB_KEY, "Authorization": f"Bearer {SB_KEY}",
        "Content-Type": "application/json"}

# ── Per-provider cost constants (USD per 1M tokens) ───────────
COST = {
    "anthropic": {
        "claude-haiku-4-5":  {"in": 0.80, "out": 4.00},
        "claude-sonnet-4-6": {"in": 3.00, "out": 15.00},
    },
    "openai": {
        "gpt-4o-mini": {"in": 0.15, "out": 0.60},
        "gpt-4o":      {"in": 2.50, "out": 10.00},
    },
}

DEFAULT_MODEL = {
    "anthropic": "claude-haiku-4-5",
    "openai":    "gpt-4o-mini",
}


class BudgetExceeded(Exception):
    """Raised when assert_budget would put the run over its cap."""


class SchemaInvalid(Exception):
    """Raised when the LLM response can't be parsed into expected shape."""


# ── Budget guard ──────────────────────────────────────────────
def _current_run_cost(run_id: str) -> float:
    r = httpx.get(
        f"{SB_URL}/rest/v1/crawler_runs",
        headers=SB_H,
        params={"select": "cost_usd", "id": f"eq.{run_id}"},
        timeout=10,
    )
    if r.status_code != 200:
        return 0.0
    rows = r.json()
    if not rows:
        return 0.0
    return float(rows[0].get("cost_usd") or 0.0)


def _add_run_cost(run_id: str, delta: float, in_tok: int, out_tok: int) -> None:
    """Increment cost_usd / tokens_in / tokens_out for the run row."""
    current = _current_run_cost(run_id)
    # NOTE: not atomic. For higher concurrency we'd want a Postgres function
    # with `UPDATE ... SET cost_usd = cost_usd + $1`. Adequate for sequential
    # single-process crawlers.
    cur_r = httpx.get(
        f"{SB_URL}/rest/v1/crawler_runs",
        headers=SB_H,
        params={"select": "tokens_in,tokens_out", "id": f"eq.{run_id}"},
        timeout=10,
    )
    cur = cur_r.json()[0] if cur_r.status_code == 200 and cur_r.json() else {}
    body = {
        "cost_usd":   round(current + delta, 4),
        "tokens_in":  (cur.get("tokens_in") or 0) + in_tok,
        "tokens_out": (cur.get("tokens_out") or 0) + out_tok,
    }
    httpx.patch(
        f"{SB_URL}/rest/v1/crawler_runs?id=eq.{run_id}",
        headers={**SB_H, "Prefer": "return=minimal"},
        json=body, timeout=10,
    )


def assert_budget(run_id: str, estimated_cost: float, max_usd_per_run: float) -> None:
    """Raise BudgetExceeded if the run would cross its cap with this call."""
    if max_usd_per_run <= 0:
        return  # 0 = unlimited
    spent = _current_run_cost(run_id)
    if spent + estimated_cost > max_usd_per_run:
        raise BudgetExceeded(
            f"run {run_id}: would spend ${spent + estimated_cost:.4f}, "
            f"cap is ${max_usd_per_run:.2f}"
        )


# ── Providers ─────────────────────────────────────────────────
def _call_anthropic(prompt: str, model: str, max_tokens: int) -> tuple[str, int, int]:
    from anthropic import Anthropic
    client = Anthropic(api_key=os.environ["ANTHROPIC_API_KEY"])
    r = client.messages.create(
        model=model,
        max_tokens=max_tokens,
        messages=[{"role": "user", "content": prompt}],
    )
    text = r.content[0].text if r.content else ""
    return text, (r.usage.input_tokens if r.usage else 0), (r.usage.output_tokens if r.usage else 0)


def _call_openai(prompt: str, model: str, max_tokens: int) -> tuple[str, int, int]:
    from openai import OpenAI
    client = OpenAI(api_key=os.environ["OPENAI_API_KEY"])
    r = client.chat.completions.create(
        model=model,
        max_tokens=max_tokens,
        response_format={"type": "json_object"},
        messages=[
            {"role": "system", "content": "Reply only with valid JSON. No prose, no markdown fences."},
            {"role": "user", "content": prompt},
        ],
    )
    text = r.choices[0].message.content or ""
    return text, (r.usage.prompt_tokens if r.usage else 0), (r.usage.completion_tokens if r.usage else 0)


# ── Public API ────────────────────────────────────────────────
def extract_json(
    *,
    prompt: str,
    run_id: str,
    max_usd_per_run: float,
    provider: str = "openai",
    model: Optional[str] = None,
    max_tokens: int = 2000,
    expected_keys: Iterable[str] = (),
    estimated_cost: float = 0.01,
) -> dict:
    """
    Call the chosen provider for structured JSON extraction.

    Raises:
      BudgetExceeded  — pre-call check failed
      SchemaInvalid   — response wasn't valid JSON, or missed expected_keys
      KeyError        — missing API key env var
    """
    if provider not in COST:
        raise ValueError(f"unknown provider: {provider}")
    model = model or DEFAULT_MODEL[provider]
    if model not in COST[provider]:
        raise ValueError(f"unknown model for {provider}: {model}")

    assert_budget(run_id, estimated_cost, max_usd_per_run)

    if provider == "anthropic":
        text, in_tok, out_tok = _call_anthropic(prompt, model, max_tokens)
    else:
        text, in_tok, out_tok = _call_openai(prompt, model, max_tokens)

    rate = COST[provider][model]
    cost = (in_tok * rate["in"] + out_tok * rate["out"]) / 1_000_000.0
    _add_run_cost(run_id, cost, in_tok, out_tok)

    # Parse — strip code fences if any, then find first JSON object
    cleaned = re.sub(r"^```(?:json)?|```$", "", text.strip(), flags=re.MULTILINE).strip()
    m = re.search(r"\{[\s\S]*\}", cleaned)
    if not m:
        raise SchemaInvalid(f"no JSON object in response: {text[:200]}")
    try:
        data = json.loads(m.group(0))
    except json.JSONDecodeError as e:
        raise SchemaInvalid(f"JSON parse failed: {e}")

    for key in expected_keys:
        if key not in data:
            raise SchemaInvalid(f"missing expected key '{key}'")

    return data

File: lib/ai/extract.ts
import Anthropic from "@anthropic-ai/sdk"
import { adminSupabase } from "@/lib/supabase"

// Provider-agnostic structured-extraction wrapper for API routes.
//
// Routes JSON-extraction to Anthropic Haiku or OpenAI gpt-4o-mini, enforces
// per-run budget caps via crawler_runs.cost_usd, schema-validates output,
// and records actual cost back to the run row.
//
// Mirrors crawlers/ai/extract.py so the budget-tracking semantics are
// identical across server-side TS and Python crawler workers.

type Provider = "anthropic" | "openai"

interface CostRate { in: number; out: number }  // USD per 1M tokens

const COST: Record<Provider, Record<string, CostRate>> = {
  anthropic: {
    "claude-haiku-4-5":  { in: 0.80, out: 4.00 },
    "claude-sonnet-4-6": { in: 3.00, out: 15.00 },
  },
  openai: {
    "gpt-4o-mini": { in: 0.15, out: 0.60 },
    "gpt-4o":      { in: 2.50, out: 10.00 },
  },
}

const DEFAULT_MODEL: Record<Provider, string> = {
  anthropic: "claude-haiku-4-5",
  openai:    "gpt-4o-mini",
}

// Default provider is Anthropic — the openai package is optional and only
// installed if you explicitly want to route Discoverer traffic to OpenAI.
// To enable: `npm install openai` and set provider: "openai" on the call.
export const DEFAULT_PROVIDER: Provider = "anthropic"

export class BudgetExceeded extends Error {
  constructor(message: string) { super(message); this.name = "BudgetExceeded" }
}

export class SchemaInvalid extends Error {
  constructor(message: string) { super(message); this.name = "SchemaInvalid" }
}

interface ExtractOpts {
  prompt: string
  runId: string
  maxUsdPerRun: number
  provider?: Provider
  model?: string
  maxTokens?: number
  expectedKeys?: string[]
  estimatedCost?: number
}

async function getCurrentRunCost(runId: string): Promise<number> {
  const { data } = await adminSupabase
    .from("crawler_runs")
    .select("cost_usd")
    .eq("id", runId)
    .maybeSingle()
  return Number((data as any)?.cost_usd ?? 0)
}

async function addRunCost(runId: string, deltaUsd: number, tokIn: number, tokOut: number): Promise<void> {
  const { data } = await adminSupabase
    .from("crawler_runs")
    .select("cost_usd, tokens_in, tokens_out")
    .eq("id", runId)
    .maybeSingle()
  const row = (data as any) ?? {}
  await adminSupabase.from("crawler_runs").update({
    cost_usd:   Number((Number(row.cost_usd ?? 0) + deltaUsd).toFixed(4)),
    tokens_in:  (row.tokens_in  ?? 0) + tokIn,
    tokens_out: (row.tokens_out ?? 0) + tokOut,
  }).eq("id", runId)
}

export async function assertBudget(runId: string, estimatedCost: number, maxUsdPerRun: number): Promise<void> {
  if (maxUsdPerRun <= 0) return  // 0 = unlimited
  const spent = await getCurrentRunCost(runId)
  if (spent + estimatedCost > maxUsdPerRun) {
    throw new BudgetExceeded(
      `run ${runId}: would spend $${(spent + estimatedCost).toFixed(4)}, cap is $${maxUsdPerRun.toFixed(2)}`
    )
  }
}

async function callAnthropic(prompt: string, model: string, maxTokens: number): Promise<{ text: string; tokIn: number; tokOut: number }> {
  const client = new Anthropic({ apiKey: process.env.ANTHROPIC_API_KEY! })
  const r = await client.messages.create({
    model,
    max_tokens: maxTokens,
    messages: [{ role: "user", content: prompt }],
  })
  const text = r.content[0]?.type === "text" ? r.content[0].text : ""
  return { text, tokIn: r.usage?.input_tokens ?? 0, tokOut: r.usage?.output_tokens ?? 0 }
}

async function callOpenAI(prompt: string, model: string, maxTokens: number): Promise<{ text: string; tokIn: number; tokOut: number }> {
  // OpenAI is optional — only used if explicitly opted into via provider: "openai".
  // We use a dynamic require() wrapped in try/catch so the build doesn't fail
  // when the openai package isn't installed.
  let OpenAI: any
  try {
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    const mod = require("openai")
    OpenAI = mod.default || mod.OpenAI || mod
  } catch {
    throw new Error(
      "OpenAI provider requested but 'openai' package is not installed. " +
      "Run 'npm install openai' or switch to provider: 'anthropic'."
    )
  }
  const client = new OpenAI({ apiKey: process.env.OPENAI_API_KEY! })
  const r = await client.chat.completions.create({
    model,
    max_tokens: maxTokens,
    response_format: { type: "json_object" },
    messages: [
      { role: "system", content: "Reply only with valid JSON. No prose, no markdown fences." },
      { role: "user", content: prompt },
    ],
  })
  return {
    text:   r.choices[0]?.message?.content ?? "",
    tokIn:  r.usage?.prompt_tokens     ?? 0,
    tokOut: r.usage?.completion_tokens ?? 0,
  }
}

export async function extractJson<T = any>(opts: ExtractOpts): Promise<T> {
  const provider     = opts.provider     ?? DEFAULT_PROVIDER
  const model        = opts.model        ?? DEFAULT_MODEL[provider]
  const maxTokens    = opts.maxTokens    ?? 2000
  const estimated    = opts.estimatedCost ?? 0.01
  const expectedKeys = opts.expectedKeys ?? []

  if (!COST[provider]?.[model]) throw new Error(`unknown ${provider} model: ${model}`)

  await assertBudget(opts.runId, estimated, opts.maxUsdPerRun)

  const { text, tokIn, tokOut } =
    provider === "anthropic"
      ? await callAnthropic(opts.prompt, model, maxTokens)
      : await callOpenAI(opts.prompt, model, maxTokens)

  const rate = COST[provider][model]
  const cost = (tokIn * rate.in + tokOut * rate.out) / 1_000_000
  await addRunCost(opts.runId, cost, tokIn, tokOut)

  const cleaned = text.replace(/^```(?:json)?\s*|\s*```$/gm, "").trim()
  const m = cleaned.match(/\{[\s\S]*\}/)
  if (!m) throw new SchemaInvalid(`no JSON object in response: ${text.slice(0, 200)}`)
  let data: any
  try { data = JSON.parse(m[0]) }
  catch (e: any) { throw new SchemaInvalid(`JSON parse failed: ${e.message}`) }
  for (const k of expectedKeys) {
    if (!(k in data)) throw new SchemaInvalid(`missing expected key '${k}'`)
  }
  return data as T
}
